go/arrow/compute/registry.go

```diff
@@ -0,0 +1,266 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package compute
+
+import (
+	"sync"
+
+	"github.com/apache/arrow/go/v10/arrow/internal/debug"
+	"golang.org/x/exp/maps"
+	"golang.org/x/exp/slices"
+)
+
+type FunctionOptionsType interface {
```

Comment:
FWIW I think we only did this for C++ metaprogramming reasons. Is there a more Go-like way to do this? For instance Copy() could be a regular method and Compare() could become Equal()?

Reply:
Hmm, OptionsTypes get registered in the registry, but looking at things, I don't actually see a direct tie of `Function`s -> their options types other than by the name (unless i'm missing something). So what is the actual use for registering the Options types in the Function Registry in the C++ Compute lib. You're right that I could place `Copy` and `Compare` as part of the `FunctionOptions` interface (as `Clone` and `Equals`) rather than needing a separate `FunctionOptionsType` interface. I think i'll do that

Reply:
It was to allow metaprogramming/serialization so you may still need an OptionsType or something like that (or maybe not depending on how those get implemented) but yeah, I would be surprised if you needed the exact same interface in Golang as C++.